### PR TITLE
TINY-5953: Autosave now more accurate on empty detection.

### PR DESCRIPTION
**Pull request by HAFRMO**
Description of Changes:
Now reuses element detection.

Pre-checks:
* [X] Changelog entry added
* [X] Tests have been added (if applicable)
* [X] Branch prefixed with `feature/` for new features (if applicable)
* [X] License headers added on new files (if applicable)

Review:
* [X] Milestone set
* [x] Review comments resolved

GitHub issues (if applicable):
Fixes https://github.com/tinymce/tinymce/issues/5328

### CODE REVIEW

File: modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts [hunk 1]
```diff
@@ -13,13 +13,18 @@ import * as Events from '../api/Events';
...
-  ).test(html);
+  if(typeof html === 'undefined') {
+    return editor.dom.isEmpty(editor.getBody());
+  } else {
+    html = Tools.trim(html);
+
+    if(html === '') {
+      return true;
+    } else {
+      const fragment = new DOMParser().parseFromString(html, 'text/html');
```
Comment (lnewson) [inline modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts:24]: Missing import here, as you need `import { DOMParser } from '@ephox/dom-globals';` to use this.

File: modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts [hunk 2]
```diff
@@ -13,13 +13,18 @@ import * as Events from '../api/Events';
...
+      return editor.dom.isEmpty(fragment as any);
```
Comment (lnewson) [inline modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts:25]: No casting as `any` here please (either fix why this was needed, or at the very least use a more specific type than any). Also, why did you need to cast this as `any` anyways?

File: modules/tinymce/src/plugins/autosave/test/ts/browser/AutoSavePluginTest.ts [hunk 3]
```diff
@@ -12,78 +12,83 @@ UnitTest.asynctest('browser.tinymce.plugins.autosave.AutoSavePluginTest', (succe
   Plugin();
   Theme();
 
-  suite.test('TestCase-TBA: AutoSave: isEmpty true', function (editor) {
-    LegacyUnit.equal(editor.plugins.autosave.isEmpty(''), true);
-    LegacyUnit.equal(editor.plugins.autosave.isEmpty('   '), true);
-    LegacyUnit.equal(editor.plugins.autosave.isEmpty('\t\t\t'), true);
+  const checkIfEmpty = (editor: Editor, html: string, isEmpty: boolean): void => {
+    Assert.eq('The HTML "' + html + '" is ' + isEmpty ? 'empty.' : 'not empty.', editor.plugins.autosave.isEmpty(html), isEmpty);
```
Comment (lnewson) [inline modules/tinymce/src/plugins/autosave/test/ts/browser/AutoSavePluginTest.ts:16]: This ternary will not work as expected, as it'll evaluate as one of the following:
- `'The HTML "' + html + '" is empty'`
- `'not empty'`

Comment (lnewson) [reply modules/tinymce/src/plugins/autosave/test/ts/browser/AutoSavePluginTest.ts:16]: This also looks to be why the DeepScan checks are failing.

File: modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts [hunk 4]
```diff
@@ -13,13 +13,18 @@ import * as Events from '../api/Events';
...
 
 const isEmpty = (editor: Editor, html?: string) => {
-  const forcedRootBlockName = editor.settings.forced_root_block;
-
-  html = Tools.trim(typeof html === 'undefined' ? editor.getBody().innerHTML : html);
-
-  return html === '' || new RegExp(
-    '^<' + forcedRootBlockName + '[^>]*>((\u00a0|&nbsp;|[ \t]|<br[^>]*>)+?|)<\/' + forcedRootBlockName + '>|<br>$', 'i'
-  ).test(html);
+  if(typeof html === 'undefined') {
```
[the remark this answers is not in the document]
Comment (lnewson) [reply modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts:16]: Also, there's a missing space after `if`.

[the remark this answers is not in the document]
Comment (HAFRMO) [reply modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts:16]: Isn't there a lintrule that'd help with the spaces for the future?

[the remark this answers is not in the document]
Comment (lnewson) [reply modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts:16]: We did have one when we were using `tslint`. A lot of things broke in the move to `eslint` it seems :disappointed:

File: modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts (hunk 2 above)
Comment (HAFRMO) [reply modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts:25]: Because DOMParser (Katamari) was not DOMParser (default), so when using DOMParser (default) it complained due to not being DOMParser (Katamari), but still worked just fine.

Comment (lnewson) [reply modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts:25]: I'm confused... there is no `DOMParser` module in `katamari`.

File: modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (HAFRMO) [reply modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts]: Ah, dammit.

File: modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts (hunk 2 above)
Comment (HAFRMO) [reply modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts]: Err, yes, dom-globals. I was still having Type on my mind when I wrote and mixed up the modules. The rest was correct though.

**Comment by metricjs** [comment]
Did you need to re-request Lee's review on this one?

File: modules/tinymce/changelog.txt [hunk 5]
```diff
@@ -7,6 +7,7 @@ Version 5.4.0 (TBD)
     Added additional font related Oxide variables for secondary buttons allowing more specific styling. #TINY-6061
     Changed `advlist` toolbar buttons to only show a dropdown list if there's more than one option #TINY-3194
     Changed `mceInsertTable` command and `insertTable` API method to take optional header rows and columns arguments #TINY-6012
+    Fixed `autosave` isEmpty API aligned further with document functions #TINY-5953
```
Comment (lnewson) [inline modules/tinymce/changelog.txt:10]: This isn't very clear, and an end-user isn't going to know what "document functions" mean. So maybe something like this instead?
```
Fixed the `autosave` isEmpty API incorrectly detecting non-empty content as empty #TINY-5953
```

File: modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts [hunk 6]
```diff
@@ -11,15 +11,22 @@ import LocalStorage from 'tinymce/core/api/util/LocalStorage';
...
-
-  html = Tools.trim(typeof html === 'undefined' ? editor.getBody().innerHTML : html);
-
-  return html === '' || new RegExp(
-    '^<' + forcedRootBlockName + '[^>]*>((\u00a0|&nbsp;|[ \t]|<br[^>]*>)+?|)<\/' + forcedRootBlockName + '>|<br>$', 'i'
-  ).test(html);
+  if (Type.isUndefined(html)) {
+    return editor.dom.isEmpty(editor.getBody());
+  } else {
+    html = Tools.trim(html);
```
Comment (lnewson) [inline modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts:21]: No mutation of variables please (ie create a new one instead).

Comment (HAFRMO) [reply modules/tinymce/src/plugins/autosave/main/ts/core/Storage.ts:21]: Should have caught that when I moved it from the old, my bad.
